App/Logic/AnalyticsLogic.swift

```diff
@@ -0,0 +1,459 @@
+      try context.awaitDispatch(SendRequest(kind: .dummy))
+    }
+  }
+
+  struct UpdateDummyTrafficOpportunityWindowIfCurrent: AppSideEffect {
+    func sideEffect(_ context: SideEffectContext<AppState, AppDependencies>) throws {
+      let state = context.getState()
+      guard state.analytics.dummyTrafficOpportunityWindow.contains(context.dependencies.now()) else {
+        // The dummy traffic opportunity is not currently open. Nothing to do
+        return
+      }
+
+      try context.awaitDispatch(UpdateDummyTrafficOpportunityWindow())
+    }
+  }
+
+  struct UpdateDummyTrafficOpportunityWindowIfExpired: AppSideEffect {
+    func sideEffect(_ context: SideEffectContext<AppState, AppDependencies>) throws {
+      let state = context.getState()
+      guard !state.analytics.dummyTrafficOpportunityWindow.contains(context.dependencies.now()) else {
+        // The dummy traffic opportunity is currently open. Nothing to do
+        return
+      }
+
+      try context.awaitDispatch(UpdateDummyTrafficOpportunityWindow())
+    }
+  }
+
+  /// Updates the dummy analytics traffic opportunity window taking the parameters from the Configuration and the RNGs.
+  struct UpdateDummyTrafficOpportunityWindow: AppSideEffect {
+    func sideEffect(_ context: SideEffectContext<AppState, AppDependencies>) throws {
+      let state = context.getState()
+
+      let dummyTrafficStochasticDelay = context.dependencies.exponentialDistributionGenerator
+        .exponentialRandom(with: state.configuration.dummyAnalyticsMeanStochasticDelay)
+
+      try context.awaitDispatch(
+        SetDummyTrafficOpportunityWindow(
+          dummyTrafficStochasticDelay: dummyTrafficStochasticDelay,
+          now: context.dependencies.now()
+        )
+      )
+    }
+  }
+}
+
+// MARK: - Token management
+
+extension Logic.Analytics {
+  struct RefreshAnalyticsTokenIfNeeded: AppSideEffect {
+    func sideEffect(_ context: SideEffectContext<AppState, AppDependencies>) throws {
+      let state = context.getState()
+      let now = context.dependencies.now()
+
+      guard let token = state.analytics.token else {
+        // No token
+        try context.awaitDispatch(RefreshAnalyticsToken())
+        return
+      }
+
+      guard !token.isExpired(now: now) else {
+        // Token expired
+        try context.awaitDispatch(RefreshAnalyticsToken())
+        return
+      }
+
+      switch token.status {
+      case .validated:
+        // Token not expired and validated
+        return
+      case .generated:
+        // Token not expired but not validated
+        try context.awaitDispatch(ValidateAnalyticsToken(token: token))
+      }
+    }
+  }
+
+  struct RefreshAnalyticsToken: AppSideEffect {
+    func sideEffect(_ context: SideEffectContext<AppState, AppDependencies>) throws {
+      let tokenString = context.dependencies.analyticsTokenGenerator.generateToken(length: AnalyticsState.AnalyticsToken.tokenLength)
```

Comment:
<!--
  0 failure: 
  1 warning:  swiftlint (0.39.2...
  
  
  DangerID: danger-id-Danger;
  File: App/Logic/AnalyticsLogic.swift;
  Line: 307;
-->

- :warning: swiftlint (0.39.2): Line should be 130 characters or less: currently 133 characters (line_length)